src/main/java/org/radarbase/management/web/rest/GroupResource.java

```diff
@@ -105,4 +113,62 @@ public GroupDTO getGroup(
         groupService.deleteGroup(projectName, groupName);
         return ResponseEntity.noContent().build();
     }
+
+    /**
+     * Add subjects to a single group.
+     * @param projectName project name
+     * @param groupName group name
+     * @param patchOperations json-patch request body
+     * @throws NotAuthorizedException if PROJECT_UPDATE permissions are not present.
+     */
+    @PatchMapping("/{groupName:" + Constants.ENTITY_ID_REGEX + "}/subjects")
+    public ResponseEntity<?> changeGroupSubjects(
+            @PathVariable String projectName,
+            @PathVariable String groupName,
+            @RequestBody List<GroupPatchOperation> patchOperations) throws NotAuthorizedException {
+        // Technically, this request modifies subjects,
+        // so it would make sense to check permissions per subject,
+        // but I assume that only those who are authorized to perform project-wide actions
+        // should be allowed to use this endpoint
+        checkPermissionOnProject(token, SUBJECT_UPDATE, projectName);
+        
+        List<GroupPatchOperation.SubjectPatchValue> addedItems = new ArrayList<>();
+        List<GroupPatchOperation.SubjectPatchValue> removedItems = new ArrayList<>();
+        for (GroupPatchOperation operation : patchOperations) {
+            String opCode = operation.getOp();
+            switch (opCode) {
+                case "add": addedItems.addAll(operation.getValue()); break;
+                case "remove": removedItems.addAll(operation.getValue()); break;
+                default:
+                    throw new BadRequestException(
+                            "Group patch operation '" + opCode + "' is not supported",
+                            GROUP, ERR_VALIDATION);
+            }
+        }
+
+        // Each item should specify either a login or an ID,
+        // since having both will require an extra validation step
+        // to reject e.g. {id: 1, login: "subject-id-42"}.
+        // Whether the IDs and logins exist and belong to the project
+        // should be checked later
+        List<GroupPatchOperation.SubjectPatchValue> allItems = new ArrayList<>();
+        allItems.addAll(addedItems);
+        allItems.addAll(removedItems);
+        for (GroupPatchOperation.SubjectPatchValue item : allItems) {
+            if (item.getId() == null && item.getLogin() == null) {
+                throw new BadRequestException(
+                    "Subject identification must be specified",
+                    GROUP, ERR_VALIDATION);
+            }
+            if (item.getId() != null && item.getLogin() != null) {
+                throw new BadRequestException(
+                    "Subject identification must be specify either ID or Login. " +
+                        "Do not provide both values to avoid potential confusion.",
+                    GROUP, ERR_VALIDATION);
+            }
+        }
```

Comment:
This could just be a function in the service and combine this code with the filtering code. Then no additional `allItems` list is needed, you just call
```java
parseSubjectSpecifications(subjectsToAdd, idsToAddList, loginsToAddList);
parseSubjectSpecifications(subjectsToRemove, idsToRemoveList, loginsToRemoveList);
```
in the service, with `idsToAddList`, etc. are empty array lists.

Reply:
If you still want to do it with streams, you could partition using collect with a map
```
partitioning = subjectsToAdd.stream()
    .collect(Collectors.groupingBy(s -> {
        Long id = s.getId();
        String login = s.getLogin();
        if (login != null && id != null) {
          return "duplicate";
        } else if (login != null) {
          return "login";
        } else if (id != null) {
          return "id";
        } else {
          return "empty";
        }
    });
if (partitioning.get("empty") != null) // bad request
if (partitioning.get("duplicate") != null) // bad request
partitioning.get("login").stream().map(::getLogin).collect(toList());
```
but it looks much more complicated this way.

Reply:
Parsing, data fetching and validation combined in `getSubjectEntities`. Now `updateGroupSubjects` looks much cleaner.